smartsim/_core/entrypoints/telemetrymonitor.py

```diff
@@ -55,17 +60,520 @@
+            collectors.append(DBConnectionCountCollector(entity, FileSink(num_out)))
+    else:
+        logger.debug(f"Collectors disabled for db {entity.name}")
+
+    return collectors
+
+
+class TaskStatusHandler(PatternMatchingEventHandler):
+    """A file listener that will notify a set of collectors when an
+    unmanaged entity has completed"""
+
+    def __init__(
+        self,
+        collector: Collector,
+        patterns: t.Optional[t.List[str]] = None,
+        ignore_patterns: t.Optional[t.List[str]] = None,
+        ignore_directories: bool = False,
+        case_sensitive: bool = False,
+    ):
+        self._collectors = [collector]
+        super().__init__(
+            patterns, ignore_patterns, ignore_directories, case_sensitive
+        )  # type: ignore
+
+    def add(self, collector: Collector) -> None:
+        self._collectors.append(collector)
+
+    def on_modified(self, event: FileSystemEvent) -> None:
+        """Event handler for when a file or directory is modified.
+
+        :param event: Event representing file/directory modification.
+        :type event: FileModifiedEvent
+        """
+        super().on_modified(event)
+        self._notify(event.src_path)
+
+    def on_created(self, event: FileSystemEvent) -> None:
+        """Event handler for when a file or directory is created.
+
+        :param event: Event representing file/directory creation.
+        :type event: FileCreatedEvent
+        """
+        super().on_created(event)
+        self._notify(event.src_path)
+
+    def _notify(self, event_src: str) -> None:
+        """Notify the collector that the entity has stopped"""
+        raise NotImplementedError("_notify is not implemented")
+
+
+class TaskCompleteHandler(TaskStatusHandler):
+    def _notify(self, event_src: str) -> None:
+        """Notify the collector that the entity has stopped"""
+        logger.debug(f"Processing stop event created @ {event_src}")
+
+        working_set = [item for item in self._collectors if item.enabled]
+        for col in working_set:
+            logger.debug(f"Disabling {col.entity.name}::{type(col).__name__}")
+            col.disable()
+            col.entity.set_complete()
+
+
+class _Address:
+    """Helper class to hold and pretty-print connection details"""
+
+    def __init__(self, host: str, port: int) -> None:
+        """Initialize the instance"""
+        self.host = host.strip() if host else ""
+        self.port = port
+        self._check()
+
+    def _check(self) -> None:
+        """Validate input arguments"""
+        if not self.host:
+            raise ValueError("Address requires host")
+        if not self.port:
+            raise ValueError("Address requires port")
+
+    def __str__(self) -> str:
+        """Pretty-print the instance"""
+        return f"{self.host}:{self.port}"
+
+
+class DBCollector(Collector):
+    """A base class for collectors that retrieve statistics from an orchestrator"""
```

Comment:
Also this implies that we have one client per metric that we're collecting. Are we potentially setting ourselves up where we start scaling like [nmetrics x nshards]. I don't really mind this, but I'm wondering if we should be sanity checking the maximum-clients parameter for the Redis cluster.

Reply:
- Per prior convo, the connections are per-node in current telmon implementation.
- Per prior convo, conn limit of `10,000` should be sufficient, even w/poor connection management

**NOTE:** ticket for follow-up comment on connection collectors will partially mitigate this issue. Planned sharing of db connections between collectors will avoid the cross-product instance problem by reducing count to `nshards` instead of `nmetrics x nshards`